### PR TITLE
test(api): Add tests for EVM emulator – API

### PR DESCRIPTION
**Pull request by slowli**
## What ❔

- Adds unit tests in the API server crate testing a mock EVM emulator.
- Allows `to == None` for `eth_call` and `debug_traceCall` RPC methods if EVM emulation is enabled, to align with Ethereum node behavior.
- Fixes an integer overflow when estimating gas for L1 / upgrade transactions.

## Why ❔

Ensures that EVM emulation will work as expected.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk_supervisor fmt` and `zk_supervisor lint`.

fix(api): Allow `to == None` for `eth_call` and `debug_traceCall`
fix(api): Avoid integer overflow when estimating gas for L1 / upgrade transactions

### CODE REVIEW

File: core/node/api_server/src/tx_sender/mod.rs [hunk 1]
```diff
@@ -102,15 +104,28 @@ impl SandboxExecutorOptions {
...
         validation_computational_gas_limit: u32,
     ) -> anyhow::Result<Self> {
+        let estimate_gas_contracts =
+            tokio::task::spawn_blocking(MultiVMBaseSystemContracts::load_estimate_gas_blocking)
+                .await
+                .context("failed loading base contracts for gas estimation")?;
+        let call_contracts =
+            tokio::task::spawn_blocking(MultiVMBaseSystemContracts::load_eth_call_blocking)
+                .await
+                .context("failed loading base contracts for gas estimation")?;
```
Comment (0xVolosnikov) [inline core/node/api_server/src/tx_sender/mod.rs:114]: Invalid text?

File: core/node/api_server/src/web3/namespaces/debug.rs [hunk 2]
```diff
@@ -262,7 +262,7 @@ impl DebugNamespace {
         let call = L2Tx::from_request(
             request.into(),
             MAX_ENCODED_TX_SIZE,
-            false, // Even with EVM emulation enabled, calls must specify `to` field
+            block_args.use_evm_emulator(),
```
Comment (0xVolosnikov) [inline core/node/api_server/src/web3/namespaces/debug.rs:265]: It is changing API, not only adding tests

Comment (slowli) [reply core/node/api_server/src/web3/namespaces/debug.rs:265]: Yep. Changed the PR description / title to reflect this.

File: core/node/api_server/src/tx_sender/gas_estimation.rs [hunk 3]
```diff
@@ -131,10 +131,7 @@ impl TxSender {
 
         if let Some(pivot) = initial_pivot {
             let iteration_started_at = Instant::now();
-            let (result, _) = estimator
-                .step(pivot)
-                .await
-                .context("estimate_gas step failed")?;
+            let (result, _) = estimator.step(pivot).await?;
```
Comment (0xVolosnikov) [inline core/node/api_server/src/tx_sender/gas_estimation.rs:134]: Why?

File: core/node/api_server/src/tx_sender/gas_estimation.rs [hunk 4]
```diff
@@ -441,22 +432,26 @@ impl<'a> GasEstimator<'a> {
...
         match &mut tx.common_data {
             ExecuteTransactionCommon::L1(l1_common_data) => {
                 l1_common_data.gas_limit = forced_gas_limit.into();
-                let required_funds =
-                    l1_common_data.gas_limit * l1_common_data.max_fee_per_gas + tx.execute.value;
+                // Since `tx.execute.value` is supplied by the client and is not checked against the current balance (unlike for L2 transactions),
+                // we may hit an integer overflow. Ditto for protocol upgrade transactions below.
+                let required_funds = (l1_common_data.gas_limit * l1_common_data.max_fee_per_gas)
+                    .checked_add(tx.execute.value)
+                    .ok_or(SubmitTxError::InsufficientFundsForTransfer)?;
```
Comment (0xVolosnikov) [inline core/node/api_server/src/tx_sender/gas_estimation.rs:444]: nit: we have several errors with similar meaning:

https://github.com/matter-labs/zksync-era/blob/6dd486555775b3dec0c0de080999a1d73aab25ce/core/node/api_server/src/tx_sender/result.rs#L19-L20C5

https://github.com/matter-labs/zksync-era/blob/6dd486555775b3dec0c0de080999a1d73aab25ce/core/node/api_server/src/tx_sender/result.rs#L35-L36

https://github.com/matter-labs/zksync-era/blob/6dd486555775b3dec0c0de080999a1d73aab25ce/core/node/api_server/src/tx_sender/result.rs#L62-L65

I suppose that we use `InsufficientFundsForTransfer` due to extremely big `tx.execute.value` required to overflow funds calculation. But as stated in the comment, we do not check value against the current balance. This behavior maybe a little bit confusing imo

**Review by 0xVolosnikov — APPROVED** [review]

**Review by 0xVolosnikov — APPROVED** [review]

File: core/lib/vm_executor/src/oneshot/contracts.rs [hunk 5]
```diff
@@ -22,11 +65,11 @@ pub(super) struct MultiVMBaseSystemContracts {
     vm_1_5_0_small_memory: BaseSystemContracts,
     /// Contracts to be used after the 1.5.0 upgrade
     vm_1_5_0_increased_memory: BaseSystemContracts,
+    _contracts_kind: PhantomData<fn() -> C>,
```
Comment (perekopskiy) [inline core/lib/vm_executor/src/oneshot/contracts.rs:68]: dumb question: why not just `PhantomData<C>`?

**Review by perekopskiy — APPROVED** [review]

File: core/lib/vm_executor/src/oneshot/contracts.rs (hunk 5 above)
Comment (slowli) [reply core/lib/vm_executor/src/oneshot/contracts.rs:68]: Here specifically, it's defensive programming out of habit 🙃 Overall, using `PhantomData<fn() -> C>`, `PhantomData<fn(C)>` etc. determines a couple of related properties, such as the layout requirements of the generic type (`PhantomData<C>` has the same layout reqs as `C`) and [variance](https://doc.rust-lang.org/nomicon/subtyping.html#variance).

**Review by 0xVolosnikov — APPROVED** [review]

**Pull request merged**